### PR TITLE
Add support to expose digest mechanisms through the provider

### PR DESCRIPTION
**Pull request by simo5**
Add the digest provider functions.
Refactor code to use helpers from digest.c so all places use a single mapping function that will be easier to keep up to date.
Add tests.

### CODE REVIEW

File: src/digests.c [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+
+    if (dctx->session == NULL) {
+        goto done;
+    }
+
+    /* This is not really funny. OpenSSL by default assumes contexts with
+     * operations in flight can be easily duplicated, with all the
+     * cryptographic status and then both context can keep going
+     * independently. We'll try here, but on failure we just 'move' the
+     * to the new token and hope for the best */
```
Comment (neverpanic) [inline src/digests.c:183]: We just move *what* to the new token? The session?

File: src/digests.c [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+    }
+
+    ret = f->C_GetOperationState(sess, state, &state_len);
+    if (ret != CKR_OK) {
+        P11PROV_raise(dctx->provctx, ret,
+                      "Error returned by C_GetOperationState");
+        goto done;
+    }
+
+    /* FIXME: should probably cycle trhough slots and check if the slot,
```
Comment (neverpanic) [inline src/digests.c:218]: Typo: `s/trhough/through/`

**Review by neverpanic — CHANGES_REQUESTED** [review]

File: src/digests.c (hunk 1 above)
Comment (simo5) [reply src/digests.c:183]: Ah I should have written, we "move the session to the new context" ... thanks for the catch I will reword.

File: src/digests.c (hunk 2 above)
Comment (simo5) [reply src/digests.c:218]: I need to find a better spell checker, the one in the CI doesn't catch much ...

File: src/signature.c [hunk 3]
```diff
@@ -337,61 +334,45 @@ static CK_RV p11prov_rsa_sig_algid(CK_MECHANISM_TYPE digest,
...
+            CK_RV rv;
+
+            rv = p11prov_digest_get_name(mech_map[i].digest, &digest);
+            if (rv != CKR_OK) {
+                return NULL;
+            }
         }
     }
-    return "";
+    return NULL;
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/signature.c:360]: It returns the digest name for the MGF.
But I should 'return digest' on line 357 and I am not, thanks for the catch (which will also break the loop).

File: src/asymmetric_cipher.c [hunk 4]
```diff
@@ -334,59 +334,45 @@ static struct {
...
+            CK_RV rv;
+
+            rv = p11prov_digest_get_name(mfg_map[i].digest, &name);
+            if (rv != CKR_OK) {
+                return NULL;
+            }
         }
     }
-    return "";
+    return NULL;
```
[the remark this answers is not in the document]
Comment (simo5) [reply src/asymmetric_cipher.c:360]: Same as above :/

File: src/digests.c (hunk 2 above)
Comment (simo5) [reply src/digests.c]: fixed

File: src/signature.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (simo5) [reply src/signature.c:361]: fixed

File: src/asymmetric_cipher.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (simo5) [reply src/asymmetric_cipher.c:361]: fixed

**Review by neverpanic — APPROVED** [review]

**Comment by simo5** [comment]
I just realized this PR conflicts with the tests refactoring @Jakuje is doing, so I will hold on pushing until that PR is in, and will adjust this one to cope.

**Pull request merged**